File: src/main/java/mods/eln/sim/electrical/heater/DiodeHeatThermalLoad.java
package mods.eln.sim.electrical.heater;

import mods.eln.sim.IProcess;
import mods.eln.sim.thermal.ThermalLoad;
import mods.eln.sim.electrical.mna.component.Resistor;

public class DiodeHeatThermalLoad implements IProcess {

    Resistor r;
    ThermalLoad load;
    double lastR;

    public DiodeHeatThermalLoad(Resistor r, ThermalLoad load) {
        this.r = r;
        this.load = load;
        lastR = r.getR();
    }

    @Override
    public void process(double time) {
        if (r.getR() == lastR) {
            load.movePowerTo(r.getP());
        } else {
            lastR = r.getR();
        }
    }
}

File: src/main/java/mods/eln/sim/thermal/DiodeHeatThermalLoad.kt
package mods.eln.sim.thermal

import mods.eln.sim.core.IProcess
import mods.eln.sim.thermal.ThermalLoad
import mods.eln.sim.mna.passive.Resistor

class DiodeHeatThermalLoad(internal var r: Resistor, internal var load: ThermalLoad) : IProcess {
    internal var lastR: Double = 0.toDouble()

    init {
        lastR = r.r
    }

    override fun process(time: Double) {
        if (r.r == lastR) {
            load.movePowerTo(r.getPower())
        } else {
            lastR = r.r
        }
    }
}
